src/backend/distributed/deparser/ruleutils_15.c

```diff
@@ -3757,6 +3731,139 @@ get_delete_query_def(Query *query, deparse_context *context,
+	{
+		char *fragmentSchemaName = NULL;
+		char *fragmentTableName = NULL;
+
+		ExtractRangeTblExtraData(targetRte, NULL, &fragmentSchemaName, &fragmentTableName, NULL);
+
+		/* use schema and table name from the remote alias */
+		appendStringInfo(buf, "MERGE INTO %s%s",
+						 only_marker(targetRte),
+						 generate_fragment_name(fragmentSchemaName, fragmentTableName));
+
+		if(targetRte->eref != NULL)
+			appendStringInfo(buf, " %s",
+					quote_identifier(get_rtable_name(query->resultRelation, context)));
+	}
+	else
+	{
+		appendStringInfo(buf, "MERGE INTO %s%s",
+						 only_marker(targetRte),
+						 generate_relation_or_shard_name(targetRte->relid,
+										context->distrelid,
+										context->shardid, NIL));
+
+		if (targetRte->alias != NULL)
+			appendStringInfo(buf, " %s",
+							 quote_identifier(get_rtable_name(query->resultRelation, context)));
+	}
+
+	/*
+	 * Add the MERGE source relation -- USING <source>
+	 */
+	get_from_clause(query, " USING ", context);
+
+	/*
+	 * Add the MERGE ON condition
+	 */
+	Assert(query->jointree->quals != NULL);
+	{
+		appendContextKeyword(context, " ON ",
+							 -PRETTYINDENT_STD, PRETTYINDENT_STD, 1);
+		get_rule_expr(query->jointree->quals, context, false);
+	}
+
+	ListCell *actionCell = NULL;
+	foreach(actionCell, query->mergeActionList)
+	{
+		MergeAction *action = (MergeAction *) lfirst(actionCell);
+
+		/* Add WHEN [NOT] MATCHED */
+		appendContextKeyword(context, " WHEN",
+							 -PRETTYINDENT_STD, PRETTYINDENT_STD, 1);
+		appendStringInfo(buf, " %s", action->matched ? "MATCHED" : "NOT MATCHED");
+
+		/* Add optional AND <condition> */
+		if (action->qual)
+		{
+			appendContextKeyword(context, " AND ",
+							 -PRETTYINDENT_STD, PRETTYINDENT_STD, 1);
+			get_rule_expr(action->qual, context, false);
+		}
+
+		appendContextKeyword(context, " THEN",
+						 -PRETTYINDENT_STD, PRETTYINDENT_STD, 1);
+
+		switch (action->commandType)
+		{
+			case CMD_INSERT:
+			{
+				appendStringInfo(buf, " INSERT " );
+				List *strippedexprs;
+				strippedexprs = get_insert_column_names_list(action->targetList,
+										buf, context, targetRte);
+				if (strippedexprs)
+				{
+					/* Add the single-VALUES expression list */
+					appendContextKeyword(context, "VALUES (",
+										 -PRETTYINDENT_STD, PRETTYINDENT_STD, 2);
+					get_rule_list_toplevel(strippedexprs, context, false);
+					appendStringInfoChar(buf, ')');
+				}
+				else
+				{
+					/* No expressions, so it must be DEFAULT VALUES */
+					appendStringInfoString(buf, "DEFAULT VALUES");
+				}
+			}
+			break;
+
+			case CMD_UPDATE:
+				appendStringInfo(buf, " UPDATE SET " );
+				get_update_query_targetlist_def(query, action->targetList,
+									context, targetRte);
+				break;
+
+			case CMD_DELETE:
+				appendStringInfo(buf, " DELETE" );
+				break;
+
+			default:
+				elog(ERROR, "unknown action in MERGE WHEN clause");
+		}
+	}
+
+	ereport(NOTICE, (errmsg("<Deparsed MERGE query: %s>", buf->data)));
```

Comment:
would be interesting to see if we can deparse all the commands in Postgres regression test suite; https://github.com/postgres/postgres/blob/master/src/test/regress/sql/merge.sql

That should give a pretty good coverage :)

Reply:
This is what I did, I took this test, as pgmerge.sql, into our code line and converted every table to Citus local table. The output (pgmerge.out) should match (barring shard names) to the expected gold file (merge.out) from community..